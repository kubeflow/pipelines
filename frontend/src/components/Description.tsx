/*
 * Copyright 2019 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import Markdown from 'markdown-to-jsx';
import { ExternalLink } from '../atoms/ExternalLink';

function preventEventBubbling(e: React.MouseEvent): void {
  e.stopPropagation();
}

const renderExternalLink = (props: {}) => (
  <ExternalLink {...props} onClick={preventEventBubbling} />
);

const options = {
  overrides: { a: { component: renderExternalLink } },
};

const optionsForceInline = {
  ...options,
  forceInline: true,
};

export const Description: React.FC<{ description: string; forceInline?: boolean }> = ({
  description,
  forceInline,
}) => {
  return <Markdown options={forceInline ? optionsForceInline : options}>{description}</Markdown>;
};
