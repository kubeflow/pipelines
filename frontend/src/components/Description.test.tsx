/*
 * Copyright 2019 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';

import { mount } from 'enzyme';
import { Description } from './Description';

describe('Description', () => {
  describe('When in normal mode', () => {
    it('renders empty string', () => {
      const tree = mount(<Description description='' />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders pure text', () => {
      const tree = mount(<Description description='this is a line of pure text' />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders raw link', () => {
      const description = 'https://www.google.com';
      const tree = mount(<Description description={description} />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders markdown link', () => {
      const description = '[google](https://www.google.com)';
      const tree = mount(<Description description={description} />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders paragraphs', () => {
      const description = 'Paragraph 1\n' + '\n' + 'Paragraph 2';
      const tree = mount(<Description description={description} />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders markdown list as list', () => {
      const description = `
* abc
* def`;
      const tree = mount(<Description description={description} />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('When in inline mode', () => {
    it('renders paragraphs separated by space', () => {
      const description = `
Paragraph 1

Paragraph 2
`;
      const tree = mount(<Description description={description} forceInline={true} />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders pure text', () => {
      const tree = mount(
        <Description description='this is a line of pure text' forceInline={true} />,
      ).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders raw link', () => {
      const description = 'https://www.google.com';
      const tree = mount(<Description description={description} forceInline={true} />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders markdown link', () => {
      const description = '[google](https://www.google.com)';
      const tree = mount(<Description description={description} forceInline={true} />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });

    it('renders markdown list as pure text', () => {
      const description = `
* abc
* def`;
      const tree = mount(<Description description={description} forceInline={true} />).getDOMNode();
      expect(tree).toMatchSnapshot();
    });
  });
});
