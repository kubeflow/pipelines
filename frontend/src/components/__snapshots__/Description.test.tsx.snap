// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Description When in inline mode renders markdown link 1`] = `
<a
  class="link"
  href="https://www.google.com"
  rel="noopener"
  target="_blank"
>
  google
</a>
`;

exports[`Description When in inline mode renders markdown list as pure text 1`] = `
<span>
  
* abc
* def
</span>
`;

exports[`Description When in inline mode renders paragraphs separated by space 1`] = `
<span>
  
Paragraph 1

Paragraph 2

</span>
`;

exports[`Description When in inline mode renders pure text 1`] = `
<span>
  this is a line of pure text
</span>
`;

exports[`Description When in inline mode renders raw link 1`] = `
<a
  class="link"
  href="https://www.google.com"
  rel="noopener"
  target="_blank"
>
  https://www.google.com
</a>
`;

exports[`Description When in normal mode renders empty string 1`] = `<span />`;

exports[`Description When in normal mode renders markdown link 1`] = `
<a
  class="link"
  href="https://www.google.com"
  rel="noopener"
  target="_blank"
>
  google
</a>
`;

exports[`Description When in normal mode renders markdown list as list 1`] = `
<ul>
  <li>
    abc
  </li>
  <li>
    def
  </li>
</ul>
`;

exports[`Description When in normal mode renders paragraphs 1`] = `
<div>
  <p>
    Paragraph 1
  </p>
  <p>
    Paragraph 2
  </p>
</div>
`;

exports[`Description When in normal mode renders pure text 1`] = `
<span>
  this is a line of pure text
</span>
`;

exports[`Description When in normal mode renders raw link 1`] = `
<a
  class="link"
  href="https://www.google.com"
  rel="noopener"
  target="_blank"
>
  https://www.google.com
</a>
`;
